modules/redis/redis.c

```diff
@@ -246,57 +233,51 @@ redis_worker_insert(LogThreadedDestDriver *s, LogMessage *msg)
       return LTR_NOT_CONNECTED;
     }
 
-  log_template_format(self->key, msg, &self->template_options, LTZ_SEND,
-                      self->super.worker.instance.seq_num, NULL, self->key_str);
-
-  if (self->param1)
-    log_template_format(self->param1, msg, &self->template_options, LTZ_SEND,
-                        self->super.worker.instance.seq_num, NULL, self->param1_str);
-  if (self->param2)
-    log_template_format(self->param2, msg, &self->template_options, LTZ_SEND,
-                        self->super.worker.instance.seq_num, NULL, self->param2_str);
+  argv = g_malloc(argc * sizeof(char *));
+  argvlen = g_malloc(argc * sizeof(size_t));
 
   argv[0] = self->command->str;
   argvlen[0] = self->command->len;
-  argv[1] = self->key_str->str;
-  argvlen[1] = self->key_str->len;
 
-  if (self->param1)
+  for (int i = 0; i < g_list_length(self->fields); i++)
     {
-      argv[2] = self->param1_str->str;
-      argvlen[2] = self->param1_str->len;
-      argc++;
+      GString *str = g_string_sized_new(1024);
+      log_template_format(g_list_nth_data(self->fields, i), msg, &self->template_options, LTZ_SEND,
+                          self->super.worker.instance.seq_num, NULL, str);
+      argvlen[i+1] = str->len;
+      argv[i+1] = g_string_free(str, FALSE);
     }
 
-  if (self->param2)
-    {
-      argv[3] = self->param2_str->str;
-      argvlen[3] = self->param2_str->len;
-      argc++;
-    }
+  reply = redisCommandArgv(self->c, argc, (const char**)argv, argvlen);
+
+  ScratchBuffersMarker marker;
+  GString *full_command = scratch_buffers_alloc_and_mark(&marker);
 
-  reply = redisCommandArgv(self->c, argc, argv, argvlen);
+  full_command = g_string_append(full_command, argv[0]);
+
+  for (int i = 1; i < argc; i++)
+    g_string_append_printf(full_command, " %s", argv[i]);
 
   if (!reply)
     {
       msg_error("REDIS server error, suspending",
                 evt_tag_str("driver", self->super.super.super.id),
-                evt_tag_str("command", self->command->str),
-                evt_tag_str("key", self->key_str->str),
-                evt_tag_str("param1", self->param1_str->str),
-                evt_tag_str("param2", self->param2_str->str),
+                evt_tag_str("command", full_command->str),
                 evt_tag_str("error", self->c->errstr),
                 evt_tag_int("time_reopen", self->super.time_reopen));
       return LTR_ERROR;
     }
 
   msg_debug("REDIS command sent",
             evt_tag_str("driver", self->super.super.super.id),
-            evt_tag_str("command", self->command->str),
-            evt_tag_str("key", self->key_str->str),
-            evt_tag_str("param1", self->param1_str->str),
-            evt_tag_str("param2", self->param2_str->str));
+            evt_tag_str("command", full_command->str));
+
   freeReplyObject(reply);
+  for (int i = 1; i < argc; i++)
+    g_free(argv[i]);
+  g_free(argv);
+  g_free(argvlen);
+  scratch_buffers_reclaim_marked(marker);
```

Comment:
This function has early return statements. We should not leak these objects in case of errors.

Reply:
Nice catch, fixed :)